File: CSLP_work3_python/Golomb.py
import math
import os
import time

from Bitstream import Bitstream


class Golomb:
    def __init__(self, filename, m):
        self.bitstream = Bitstream(filename)
        self.m = m

    def encode(self, n, last=False):
        r = n % self.m
        q = n // self.m

        for i in range(q):  # unary side of the code
            self.bitstream.write_one_bit(1)

        self.bitstream.write_one_bit(0)

        self.bitstream.write_n_bits(r, self.m)  # binary side of the code

        if self.bitstream.bitstream and last:
            self.bitstream.padding_last_byte = 8 - len(self.bitstream.bitstream)
            self.bitstream.write_to_file()

        return self.bitstream.bitstream

    def decode(self):
        k = math.ceil(math.log2(self.m))
        t = math.pow(2, k) - self.m
        r = 0
        q = 0
        i = 0

        while True:
            #print("Hey: ", self.bitstream.bitstream)
            bit = self.bitstream.read_one_bit()
            if bit == '0':
                break
            q += 1
            i += 1

        #print("Q: ", q, self.bitstream.bitstream, k)
        temp = self.bitstream.read_n_bits(0, k - 1)
        r = int("".join(str(x) for x in temp), 2)
        #print("R: ", r, "T: ", t, self.bitstream.bitstream)

        if r < t:
            result = q * self.m + r
            #print("Resultado: ", result, self.bitstream.bitstream)
            return result

        else:
            r = r * 2 + int(self.bitstream.read_one_bit())
            result = q * self.m + r - t
            #print("Resultado: ", result, self.bitstream.bitstream)
            return result



def main():
    g = Golomb('outputfile.txt', 78)
    g.bitstream.open_file_write()
    g.encode(261)
    g.encode(250)
    g.encode(257)
    g.encode(253)
    g.encode(256)
    g.encode(1)
    g.encode(25)
    g.encode(258)
    if g.bitstream.bitstream:
        g.bitstream.padding_last_byte = 8 - len(g.bitstream.bitstream)
        g.bitstream.write_to_file()

    g.bitstream.close_file()

    g.bitstream.open_file_read()
    print(g.decode())
    print(g.decode())
    print(g.decode())
    print(g.decode())
    print(g.decode())
    print(g.decode())
    print(g.decode())
    print(g.decode())
    g.bitstream.close_file()


if __name__ == '__main__':
    main()

File: CSLP_work3_Cpp/Golomb.cpp
/*
 * Golomb.cpp
 *
 *  Created on: 21 Nov 2019
 *      Author: ghost
 */

#include "Golomb.h"
#include <functional>
#include <iostream>
#include <cmath>
#include <bitset>
#include <string>

using namespace std;



Golomb::Golomb(int m1) {
	Bitstream bt;
	m = m1;
}

Bitstream Golomb::encode(int n){

	int r = n % getM();
	int q = n / getM();

	for(int i = 0; i < q; i++){
		bt.writeOneBit(1);
	}


	bt.writeOneBit(0);
	bt.writeNBits(r, getM());


	return bt;
}

int Golomb::decode(){
	int k = ceil(log2(getM()));
	int t = pow(2, k) - getM();
	int r = 0;
	int q = 0;

	while(true){
		//cout << "here: " << bt.getBitstream() << endl;
		char bit = bt.readOneBit();
		if(bit == '0'){
			break;
		}
		q += 1;
	}
	/*
	cout << q << "\n";
	cout << k << "\n";
	*/

	std::string temp = bt.readNBits(0, k - 1);
	r = std::stoi(temp, nullptr, 2);

	/*
	cout << r << "\n";
	cout << t << "\n";
	*/

	if(r < t){
		return q * getM() + r;
	}
	else{
		r = r * 2 + (int)bt.readOneBit() - 48; // 48 vem de 0 no tabela ascii ser 48
		return q * getM() + r - t;
	}

}

Bitstream& Golomb::getBitstream(){
	return bt;
}

int Golomb::getM(){
	return m;
}


int main(){
	Golomb g(5);
	cout << g.encode(15).getBitstream() << "\n";
	cout << g.encode(4).getBitstream() << "\n";
	cout << g.encode(16).getBitstream() << "\n";
	cout << g.encode(12).getBitstream() << "\n";

	if(g.getBitstream().getBitstream().length() > 0){
		g.getBitstream().lastByteLength = g.getBitstream().getBitstream().length();
		g.getBitstream().writeToFile();
	}
	else
		g.getBitstream().lastByteLength = 8;

	g.getBitstream().readFile();
	cout << g.decode() << "\n";
	cout << "---------------------------------------" << endl;
	cout << g.decode() << "\n";
	cout << "---------------------------------------" << endl;
	cout << g.decode() << "\n";
	cout << "---------------------------------------" << endl;
	cout << g.decode() << "\n";
}




